Kernel/Storage/ATA/AHCIController.cpp

```diff
@@ -26,22 +26,37 @@ UNMAP_AFTER_INIT NonnullRefPtr<AHCIController> AHCIController::initialize(PCI::D
 
 bool AHCIController::reset()
 {
-    hba().control_regs.ghc = 1;
-
-    dbgln_if(AHCI_DEBUG, "{}: AHCI Controller reset", pci_address());
-
-    full_memory_barrier();
-    size_t retry = 0;
+    dmesgln("{}: AHCI controller reset", pci_address());
+    {
+        SpinlockLocker locker(m_hba_control_lock);
+        hba().control_regs.ghc = 1;
+
+        dbgln_if(AHCI_DEBUG, "{}: AHCI Controller reset", pci_address());
+
+        full_memory_barrier();
+        size_t retry = 0;
+
+        // Note: The HBA is locked or hung if we waited more than 1 second!
+        while (true) {
+            if (retry > 1000)
+                return false;
+            if (!(hba().control_regs.ghc & 1))
+                break;
+            IO::delay(1000);
+            retry++;
+        }
+        // Note: Turn on AHCI HBA and Global HBA Interrupts.
+        full_memory_barrier();
+        hba().control_regs.ghc = (1 << 31) | (1 << 1);
+        full_memory_barrier();
+    }
 
-    while (true) {
-        if (retry > 1000)
-            return false;
-        if (!(hba().control_regs.ghc & 1))
-            break;
-        IO::delay(1000);
-        retry++;
+    auto implemented_ports = AHCI::MaskedBitField((u32 volatile&)(hba().control_regs.pi));
```

Comment:
`m_ports` should be cleared here, if the reset resulted in a different `pi` value, we would have stale left-over ports in there

Reply:
Well, according to the AHCI spec:

> This register indicates which ports are exposed by the HBA. It is loaded by the BIOS. It indicates which
ports that the HBA supports are available for software to use. For example, on an HBA that supports 6
ports as indicated in CAP.NP, only ports 1 and 3 could be available, with ports 0, 2, 4, and 5 being
unavailable.

Which means that condition is not possible, but sure, I'll add that line with a comment about it :)